component/discovery/kubernetes/k8s.go

```diff
@@ -0,0 +1,149 @@
+func (sd *SDConfig) Convert() *promk8s.SDConfig {
+	selectors := make([]promk8s.SelectorConfig, len(sd.Selectors))
+	for i, s := range sd.Selectors {
+		selectors[i] = *s.convert()
+	}
+	return &promk8s.SDConfig{
+		APIServer:          sd.APIServer.Convert(),
+		Role:               promk8s.Role(sd.Role),
+		KubeConfig:         sd.KubeConfig,
+		HTTPClientConfig:   *sd.HTTPClientConfig.Convert(),
+		NamespaceDiscovery: *sd.NamespaceDiscovery.convert(),
+		Selectors:          selectors,
+	}
+}
+
+// NamespaceDiscovery mirroring prometheus type
+type NamespaceDiscovery struct {
+	IncludeOwnNamespace bool     `river:"own_namespace,attr,optional"`
+	Names               []string `river:"names,attr,optional"`
+}
+
+func (nd *NamespaceDiscovery) convert() *promk8s.NamespaceDiscovery {
+	return &promk8s.NamespaceDiscovery{
+		IncludeOwnNamespace: nd.IncludeOwnNamespace,
+		Names:               nd.Names,
+	}
+}
+
+// SelectorConfig mirroring prometheus type
+type SelectorConfig struct {
+	Role  string `river:"role,attr"`
+	Label string `river:"label,attr,optional"`
+	Field string `river:"field,attr,optional"`
+}
+
+func (sc *SelectorConfig) convert() *promk8s.SelectorConfig {
+	return &promk8s.SelectorConfig{
+		Role:  promk8s.Role(sc.Role),
+		Label: sc.Label,
+		Field: sc.Field,
+	}
+}
+
+// Exports holds values which are exported by the discovery.k8s component.
+type Exports struct {
+	Targets []scrape.Target `river:"targets,attr"`
+}
+
+// Component implements the discovery.k8s component.
+type Component struct {
+	opts   component.Options
+	cancel context.CancelFunc
+
+	newDiscoverer chan discovery.Discoverer
+}
+
+// New creates a new discovery.k8s component.
+func New(o component.Options, args SDConfig) (*Component, error) {
+	c := &Component{
+		opts: o,
+		// buffered to avoid deadlock from the first immediate update
+		newDiscoverer: make(chan discovery.Discoverer, 1),
+	}
+	return c, c.Update(args)
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case disc := <-c.newDiscoverer:
+			// cancel any previously running discovery
+			if c.cancel != nil {
+				c.cancel()
+			}
+			// function to send updates on change
+			f := func(t []scrape.Target) {
+				c.opts.OnStateChange(Exports{Targets: t})
+			}
+			// create new context so we can cancel it if we get any future updates
+			// since it is derived from the main run context, it only needs to be
+			// canceled directly if we receive new updates
+			newCtx, cancel := context.WithCancel(ctx)
+			c.cancel = cancel
+			// finally run discovery
+			go discovery.RunDiscovery(newCtx, disc, f)
+		}
+	}
+}
+
+// Update implements component.Compnoent.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(SDConfig)
+	disc, err := promk8s.New(c.opts.Logger, newArgs.Convert())
+	if err != nil {
+		return err
+	}
+	c.newDiscoverer <- disc
```

Comment:
Hmm. 

I think this is probably fine, but it might cause a deadlock if someone Update gets called twice when a component it still exiting. 

Just to be safe, can we change this to

```go
select {
case c.newDiscoverer <- disc: 
default: 
}
```

(Though honestly that's not ideal either; storing the latest discoverer in a field and using a notification channel might be more 'correct' but I think we need to change something here just to be safe, especially while the component controller is still young)